### ISSUE TITLE
drun shows duplicate desktop entries

### ISSUE THREAD

**Issue by fice-t** [DaveDavenport/rofi#453]
## Version

Version: 1.2.0

## Configuration
Default config.

## Launch Command
`rofi -show drun`

## Steps to reproduce

* Copy a .desktop file from `/usr/share/applications` into `~/.local/share/applications`
* Run `rofi -show drun` and search for the program whose desktop file was copied

## What behaviour you see

* - Rofi shows desktop files that appear in both `~/.local/share/applications` and `/usr/share/applications` twice

## What behaviour you expect to see

* - Desktop files in `/usr/share/applications` that are shadowed by entries in  `~/.local/share/applications` do not appear in Rofi

**Comment by carnager** [comment]
I agree, this is a bug, since the standard clearly defines, that local copies override their global counterparts.

**Comment by DaveDavenport** [comment]
@carnager can you add link to spec.

**Comment by sardemff7** [comment]
[This is](https://standards.freedesktop.org/desktop-entry-spec/latest/ape.html) the relevant part for this bug.

**Comment by DaveDavenport** [comment]
Should be fixed in git. Compile with `--enable-drun`.

**Comment by fice-t** [comment]
Unfortunately it seems that the system-wide .desktop files are now overriding the local ones.

For example, copy a desktop file from `/usr/share/applications` into `~/.local/share/applications` and replace its contents with:

```
[Desktop Entry]
Hidden=true
```
This should mean that this entry should not appear in Rofi, but it has no effect.

**Comment by DaveDavenport** [comment]
No, local file are parsed first.
However if hidden is set the file is indeed skipped, the global will then show up.  Did not realize this use-case, Need to see how to solve this one :SMILE:.

**Comment by DaveDavenport** [comment]
Added a fix, can you test?

(Keep a list of Desktop ID's that are disabled, so I can skip them later).

**Comment by DaveDavenport** [comment]
```G_MESSAGES_DEBUG=Dialogs.DRun ./rofi -show drun```

for more info.

**Comment by fice-t** [comment]
> No, local file are parsed first.

Right, I just had another issue, but it was solved by clearing the drun cache (the name of the system copy was used instead of the local name).

> Added a fix, can you test?

It doesn't work for the above case as the desktop file apparently is skipped instead of added to the disabled list:
`(process:4113): Dialogs.DRun-DEBUG: Skipping desktop file: /home/alex/.local/share/applications/emacs24-term.desktop because: No type indicated
`

I suppose per the specification a `Type` field should be there, but other applications that parse .desktop files seem to handle the above just as it is.

**Comment by DaveDavenport** [comment]
The specification specifies it is a required field. I am not going to support broken files.

**Comment by fice-t** [comment]
That's fair. Though I did take another look at the spec, which says:

> If a REQUIRED key is only valid in the context of another key set to a specific value, then it has to be present only if the other key is set to the specific value

So one could say that the `Type` key being required is only valid in the case where `Hidden=false` as the spec says for the `Hidden` keyword:

> It's strictly equivalent to the .desktop file not existing at all, as far as that user is concerned. 

Meaning that if `Hidden=true`, then `Type` does not have to appear in the file. The same goes for `Name`.

This part isn't really a big deal, but looking at drun.c it seems that all it needs is for the `Hidden` block to come before the `Type` block. It looks like it already comes before the `Name` check anyway.

**Comment by DaveDavenport** [comment]
oeps,  you are right.

**Comment by fice-t** [comment]
> and desktop-file-validate agrees with that.

Mildly disappointing, but I don't care enough to challenge that. 

Though I tested it out and it indeed is just a matter of putting the `// skip hidden entries` block above the `// skip non Application entries`.
Currently it works without a `Name` field even though it's "Required", so it wouldn't be a stretch to do the same for `Type`.

I'll close the issue since the main issue has been solved, thanks.
